### PR TITLE
addressing velocity discontinuities correctly in the cost function

### PR DESCRIPTION
**Pull request by yangwill**
Implementation of a more general version of #231.

Creates a Substitution member variable in Dircon that replaces the velocity variables with the correct post-impact velocity variables.

---
This change is [Reviewable](https://reviewable.io/reviews/dairlab/dairlib/236)

### CODE REVIEW

File: systems/trajectory_optimization/dircon/dircon.cc [hunk 1]
```diff
@@ -102,6 +103,11 @@ Dircon<T>::Dircon(std::unique_ptr<DirconModeSequence<T>> my_sequence,
     if (i_mode > 0) {
       v_post_impact_vars_.push_back(NewContinuousVariables(
           plant_.num_velocities(), "v_p[" + std::to_string(i_mode) + "]"));
+      v_post_impact_vars_substitute_.push_back(
+          {state().tail(plant_.num_velocities()),
+           Eigen::Map<MatrixXDecisionVariable>(
+               v_post_impact_vars_.back().data(), plant_.num_velocities(), 1)
+               .cast<Expression>()});
```
Comment (mposa) [inline systems/trajectory_optimization/dircon/dircon.cc:110]: Why use Eigen::Map here? An alternate option would be 
```
v_post_impact_vars_.back().tail(plant_.num_velocities()).template cast<Expression>()}
```

It's a bit easier to read, though without the memory efficiency of Map.

File: systems/trajectory_optimization/hybrid_dircon.cc [hunk 2]
```diff
@@ -107,6 +109,17 @@ HybridDircon<T>::HybridDircon(const MultibodyPlant<T>& plant,
           constraints_[i]->countConstraintsWithoutSkipping(),
           "impulse[" + std::to_string(i) + "]"));
     }
+    //  Post-impact variables. Note: impulse variables declared below.
+    if (i > 0) {
+      v_post_impact_vars_substitute_.push_back(
```
Comment (mposa) [inline systems/trajectory_optimization/hybrid_dircon.cc:114]: See comment on dircon.cc

**Review by mposa — APPROVED** [review]
Looks good to me. Just the one comment/suggestion.

File: systems/trajectory_optimization/dircon/dircon.cc (hunk 1 above)
Comment (yangwill) [reply systems/trajectory_optimization/dircon/dircon.cc:110]: I was using https://github.com/RobotLocomotion/drake/blob/a3312d0f2d526c074f793c0c7eba7f41153c7055/systems/trajectory_optimization/multiple_shooting.cc#L95 as a reference. Though I'm not confident if `template_cast` wouldn't achieve the same effect.

**Pull request merged**